sql: Restrict renames of objects that views depend on

a-robinson opened the pull request:
<!-- Reviewable:start -->

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/9988)

<!-- Reviewable:end -->

a-robinson committed:
sql: Restrict renames of objects that views depend on

a-robinson committed:
sql: Only include needed columns in a view's table references

a-robinson commented:
Also fix creation of back-references to only include columns actually used by the view query.

@dt

a-robinson commented on pkg/sql/testdata/rename_index:
```diff
@@ -74,3 +74,12 @@ SELECT * FROM users
+statement ok
+CREATE VIEW v AS SELECT name FROM users@{FORCE_INDEX=ufo}
+
+statement error cannot rename index "ufo" because it is depended on by view "test.v"
```
That'd certainly be preferable behavior, but we've basically punted on query rewriting until we can have a proper semantic encoding of the view's query rather than the current text-based one - https://github.com/cockroachdb/cockroach/pull/9045#issuecomment-245769303

Unless you're suggesting that perhaps this particular case of rewriting is simple enough to handle?

petermattis replied:
No, I wasn't suggesting the rewriting was easy, just wondering what the long-term goal is here. This is fine in the short-term.

One thought about how to avoid the VIEW query rewriting is that we can extend our SQL syntax in order to allow referring to tables/columns/indexes by ID. Then we'd do the name resolution once when the VIEW is created. Easy to describe this, but it would likely be a whole bunch of work to actually accomplish.

a-robinson committed:
sql: Improve view-related drop error messages and test coverage

a-robinson commented:
Friendly ping :)

Also, I've added a commit here to improve the test coverage for dropping indexes that a view depends on and the error messages for view-related drop errors. It's fairly independent logically, but I wanted to reuse some of the error-generating code that I added for renames to add detail to the error messages that I think improves them. Feel free to disagree on that though, perhaps the extra verbosity isn't useful?

dt commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> though we could drop TODOs in to do the query rewrite, or even use the unimplemented error message that includes the issue to help clue in users who might expect it to Just Work.

---

Reviewed 1 of 1 files at r1.
Review status: 1 of 13 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9988#-:-KUSL7YaLgpIgrbnGIR8:bje7pu5)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Good point, added a TODO. I didn't update the error message because I'm not sure of a way to phrase it that isn't pretty unwieldy, but do you think I should?
